Document enumerating tree topologies in combinatorics module

daniel-goldstein opened the pull request:
This fills out the top-level documentation for the combinatorics module. It addresses the two key pieces of functionality: ranking/unranking trees and enumerating all trees, and how the two are intimately connected. Also provides a short example of using ranks to count tree topologies in a tree sequence.

Resolves #643

jeromekelleher commented on docs/combinatorics.rst:
```diff
@@ -1,6 +1,202 @@
+=============
+Combinatorics
+=============
+This module implements a combinatorial approach to counting topologies
```
We're not implementing it as an externally visible module - can you change the intro to talking about tskit in general rather than the combinatorics module?

daniel-goldstein replied:
Do you want to keep the generating functions in the combinatorics module, or put them under a different namespace? That would affect whether the module needs to be visible.

jeromekelleher replied:
We can make them visible externally by exporting from ``__init__.py``, like 
```
from . combinatorics import all_tree_topologies 
```
 etc. (Maybe this is a better name than ``all_unlabelled_trees``?)

daniel-goldstein replied:
> We can make them visible externally by exporting from `__init__.py`, like

Sounds good (I'm flattered my functions get top-level)

> etc. (Maybe this is a better name than all_unlabelled_trees?)

I think we've been referring to topologies without further specification as the leaf-labelled trees. Based on what I have now, we could do:

```
all_tree_topologies(n) (right now this is all_trees)
all_tree_shapes(n)
all_tree_labellings(t)
```

or could switch it to something like:

```
all_labelled_tree_topologies(n)
all_tree_topologies(n)
all_tree_labellings(t)?
```

jeromekelleher commented on docs/combinatorics.rst:
```diff
@@ -1,6 +1,202 @@
++++++++++++++++++++++++++++++++++
+Ranking/Unranking Topologies
++++++++++++++++++++++++++++++++++
+The enumeration of leaf-labelled trees discussed below allows for an
```
Which trees? It feels like there used to be a sentence before this. It would be good to briefly discuss what kind of trees we're talking about, I guess.

jeromekelleher replied:
Maybe helpful here to show all the trees of 4 (or something) just as an intuitive way into things?

jeromekelleher replied:
The concept of rank should be clear then.

jeromekelleher commented on docs/combinatorics.rst:
```diff
@@ -1,6 +1,202 @@
+Enumerating Topologies
+++++++++++++++++++++++
+
+Below we outline the approach for enumerating all unique topologies
```
There's a lot of "Belows" and stuff. How about  "Enumerating all unique topologies of leaf-labelled trees with n leaves consists of two steps".

jeromekelleher commented on docs/combinatorics.rst:
```diff
@@ -1,6 +1,202 @@
+following enumeration algorithm, which is presented below in both the
+binary and general case.
+
+Examining first the binary case, let :math:`T = (T_j, T_k)`, with left subtree
```
At this point, I'm not sure that the algorithm is what we want. Probably best to keep this level of detail for a more technical writeup. I guess what we're trying to get across here in these docs is, *what* it does, not *how* it does it.

petrelharp replied:
Agree - I don't think this is an enumeration algorithm? I see what the ordering *is*, but not everything you say below comes in a natural order, as far as I can tell.

Oh, wait: actually, the ordering itself is a bit confusing, as it's defined implicitly in terms of the orderings of smaller-n labelings? You've said how to compare labelings with different labels; but to rank labelings you need to also compare labellings with the same labels. Maybe inserting a definition along the lines of the one above would be good. I guess that this is that the rank of an labeling of a tree is its rank among sorted tuples `((size_1, shape_1, labeling_rank_1), ... (size_1, shape_1, labeling_rank_1))`, where `labeling_rank_1` is the rank of that particular labeling of labels [1...n] to the subtree with shape `shape_1` and size `size_1` (which is less than `n`)?

I may not have been very clear there. I am guessing that the labeling
```
  ┃  
 ┏┻━┓
 ┃  ┃
┏┻┓ ┃
┃ ┃ ┃
1 2 0
```
comes before
```
  ┃  
 ┏┻━┓
 ┃  ┃
┏┻┓ ┃
┃ ┃ ┃
0 2 1
```
because the smallest (and thus earliest) subtree has the smaller label in the first labeling?  Is this true?

daniel-goldstein replied:
@petrelharp I *think* I understand what you're asking and yes, there is a piece missing here. The way in which labellings are created is directly dependent on how you choose combinations of labels when splitting labels up between subtrees. This uses a lexicographic ordering for choose that comes with `itertools.combinations` to select labels to assign to subtrees, so for the trees you mention it would be ordered (0, (1, 2)), (1, (0, 2)), (2, (0, 1))

petrelharp replied:
Ah-ha - that's equivalent to what I said, but simpler. So, you're saying that labelings are ordered lexicographically, basically, with the grouping of labels determined by the subtrees, and the ordering of these groups determined by subtree order as described before.

petrelharp commented on docs/combinatorics.rst:
```diff
@@ -1,6 +1,202 @@
+
+rank
+    The rank of a leaf-labelled tree is considered to be the tuple of its
+    ``shape rank`` and ``label rank``.
```
Usually, a rank is a single number, so at first I thought you meant that the rank is the rank of the tuple in the enumerations of all such tuples; but you actually mean the tuple, right? Perhaps this is worth explaining, e.g., adding "(Note: this 'rank' is a pair of numbers, not a single number, but we call it a rank because it can be used to order labeled trees.)"

petrelharp commented on docs/combinatorics.rst:
```diff
@@ -1,6 +1,202 @@
+of leaf-labelled trees with :math:`n` leaves. This takes part in two
+steps. First, we must consider how to generate unique tree shapes.
+Then, for each shape we must consider how to uniquely assign labels
+to the leaves. Note that this enumeration is not limited to binary trees.
```
".. takes part.." --> "This happens in two steps."  ?

petrelharp replied:
Also, instead of "Note that this enumeration is not limited to binary trees", maybe you should say, more explicitly, "This is an enumeration of all general rooted trees, not only binary trees." (oh: the word "rooted" is important, too)

petrelharp replied:
You should probably say early on that all your trees are rooted.

petrelharp commented on docs/combinatorics.rst:
```diff
@@ -1,6 +1,202 @@
+A tree with :math:`n > 1` leaves can be obtained by joining at least two trees whose number
+of leaves sum to :math:`n`. This maps very closely to the concept of integer partitions.
+We first compute all integer partitions* of `n`, and then for each partition :math:`P`,
+compute the cross product of all :math:`k`-leaf trees for all :math:`k \in P`.
```
instead of "cross product", how about "combinations"?

petrelharp commented on docs/combinatorics.rst:
```diff
@@ -1,6 +1,202 @@
+
+A tree with :math:`n > 1` leaves can be obtained by joining at least two trees whose number
+of leaves sum to :math:`n`. This maps very closely to the concept of integer partitions.
+We first compute all integer partitions* of `n`, and then for each partition :math:`P`,
```
shouldn't you say "all nondecreasing integer partitions", and explain what "nondecreasing" means?

petrelharp commented on docs/combinatorics.rst:
```diff
@@ -1,6 +1,202 @@
+However, this method has the potential to double-count trees when we encounter
+partitions with repeated numbers of leaves. To remove these redundant trees we
+reduce the set of trees so that if a subtree has the same number of leaves as
+its left sibling, it must have an equal or greater shape rank.
```
very nice solution!

petrelharp replied:
So, I guess that what you're doing is saying the subtrees must be ordered by `(size, rank)`? So the rank of a tree among trees with `n` leaves is the same as the rank of its ordered list of subtrees: `((size_1, rank_1), ..., (size_n, rank_n))` among all such ordered lists?

daniel-goldstein replied:
Yes, for unlabelled trees. The definition is extended when you add labels. Precisely, it's

https://github.com/tskit-dev/tskit/blob/72ac9d2472f645163c0675c4413ddc3cb6e709b6/python/tskit/combinatorics.py#L264-L269

daniel-goldstein commented:
> This is great @daniel-goldstein! I think we can keep the detailed algorithmic desciptions out of here (we're explaining _what_ not _how_).
> 
> What we're missing is references to the actual methods. I think it would be good to have a table (or something) right at the top of the page that list does something like
> 
> Tree.rank | Return the rank of a tree
> Tree.unrank | Returns a tree with a given rank
> all_labelled_trees | Returns a generator over all labelled trees with n leaves.
> 
> Where the function name is a click-through link to the documentation.
> 
> I think a short demonstration in a section at the end would be really helpful too.

Ok great, that makes sense. I'll save the algorithmic details for a technical write-up (but will incorporate the helpful feedback!)

petrelharp commented on docs/combinatorics.rst:
```diff
@@ -1,6 +1,202 @@
+then the trees generated by :math:`P` are isomorphic to :math:`S_j \times S_k`.
+If :math:`j = k`, then the trees generated are isomorphic to all 2-element
+combinations with replacement of trees from :math:`S_j`.
+This generalizes to the following algorithm run per partition:
```
Should say what this is an algorithm to do (to enumerate binary, rooted trees with n leaves, right?).  Is this something you implement? If not, maybe you don't need it, or else say that it's for illustration.

daniel-goldstein commented:
Thanks @jeromekelleher @petrelharp! I've incorporated your feedback and moved the algorithmic discussion out of this doc and saved it for later. Went short and sweet here with laying out the current functionality and briefly explaining what these weird ranks are.

daniel-goldstein commented on docs/combinatorics.rst:
```diff
@@ -1,6 +1,116 @@
+                                      tree shapes of n leaves.
+:func:`tskit.all_tree_labellings`     Return a generator over all
+                                      labellings of the given tree's shape.
+=================================     =====================================
```
@jeromekelleher Went with this naming pattern but happy to change it up if you have a better suggestion.

jeromekelleher replied:
I was thinking of making "topologies" synonymous with "tree shapes" , but I think you're probably right. How about ``all_tree_topologies`` -> ``all_trees``, and then keep the other two as they are? Usually when someone thinks of a tree in this context it'll be labelled.

petrelharp commented on docs/combinatorics.rst:
```diff
@@ -1,6 +1,116 @@
+Combinatorics
+=============
+tskit uses a combinatorial approach to identify unique topologies of
+rooted, leaf-labelled trees. It provides methods for enumerating
```
whoops - one more caveat - "rooted, leaf-labeled trees without unary nodes"?

petrelharp commented:
Gee, I'm worried I gave you the wrong idea! A bunch of that stuff you had written earlier was very good, and very helpful! It explained what exactly the rank is, which was really neat!  It just needed a bit of adjustment, I think. I'd vote to put some of it back in - at least, it'd be nice to have a concrete definition of how the rank is computed?

daniel-goldstein commented:
> Gee, I'm worried I gave you the wrong idea! A bunch of that stuff you had written earlier was very good, and very helpful! It explained what exactly the rank is, which was really neat! It just needed a bit of adjustment, I think. I'd vote to put some of it back in - at least, it'd be nice to have a concrete definition of how the rank is computed?

No worries! I figured it'd be easier to overcorrect and then add things back from my other commit, but I am having trouble finding the line at which it becomes overly technical. Maybe just a small bit about how shapes are computed from partitions (perhaps with the accompanying diagram I had) and then combinations of labels selected in a lexicographic order?

daniel-goldstein commented:
@petrelharp Added a commit with what I think are the important bits about enumerating shapes and labellings that contribute to a complete ordering of topologies.

jeromekelleher commented on docs/combinatorics.rst:
```diff
@@ -1,6 +1,116 @@
+
+::
+
+   Ranks of 3-leaf trees: [(0, 0), (1, 0), (1, 1), (1, 2)]
```
Out of interest, is there a bijection of these tuples to integers? It would be nicer if we could just define the rank as an integer.

jeromekelleher replied:
It would be *super* nice if the rank was a direct index into ``list(all_trees(n))``

daniel-goldstein replied:
_This_ has been bugging me for a month. A naive bijection would essentially be:
```
sum(num_labellings(s, n) for s in range(shape_rank)) + label_rank
```

but I have not thought of a way yet without running through all preceding shapes, because depending on the number of symmetries shapes can be labelled in varying numbers of ways.

jeromekelleher replied:
OK, let's not get caught up on it. If we figure out a bijection we can change things later.

petrelharp replied:
I dunno - having a 2d index is kinda nicer because it makes it easier to separate tree shape from labeling?

daniel-goldstein replied:
How often would you want to just know about the shape? (labelling means nothing on its own)

jeromekelleher replied:
A bijection would give you both - you could move back and forth between the two representations. I can't see anyone being particularly interested in the actual rank values beyond defining the actual tree. If this rank was also its index in ``all_trees()``, then so much... neater.

jeromekelleher commented on docs/python-api.rst:
```diff
@@ -487,6 +487,18 @@ using a schema. See :ref:`sec_metadata`, :ref:`sec_metadata_api_overview` and
+*************
+Combinatorics
+*************
+The following are generators for fully enumerating unique tree topologies.
```
A link to ``:meth:`Tree.rank` and tree.unrank would be helpful here also.

daniel-goldstein commented on docs/combinatorics.rst:
```diff
@@ -1,6 +1,207 @@
+
+Note that the middle column reflects all tree shapes of three leaves
+in the right subtree!
+
```
Probably should mention here that the order of partitions is using ascending compositions?

jeromekelleher replied:
That fitted in naturally to your earlier exposition, I think (which I think we should include as another section at the end). Can link to my thesis, if someone wants to know what an ascending composition is, I guess?

jeromekelleher replied:
Sorry, I was commenting out of place here. If you like you can stick in a link to my summary page on [ascending compositions](http://jeromekelleher.net/generating-integer-partitions.html) as a reference.

daniel-goldstein replied:
Whoops sorry forgot this comment.

daniel-goldstein commented:
> I was thinking of making "topologies" synonymous with "tree shapes" , but I think you're probably right. How about all_tree_topologies -> all_trees, and then keep the other two as they are? Usually when someone thinks of a tree in this context it'll be labelled.

I should have waited. I previously had it as `all_trees` 😄 

I think we overlapped in review/new commit @jeromekelleher. To clarify, do you mean that the section I just added "Enumerating Topologies" is sufficient in place of your suggested "Formal definition"? Or are there additional details you think I should bring back?

jeromekelleher commented:
> I think we overlapped in review/new commit @jeromekelleher. To clarify, do you mean that the section I just added "Enumerating Topologies" is sufficient in place of your suggested "Formal definition"? Or are there additional details you think I should bring back?

Sorry, you're right. I was commenting on the version just before you put in the section. It looks great now. I do like ``all_trees`` though, it's neat and this is the main function that people will want to use when exploring this stuff.

daniel-goldstein commented:
Great, updated/squashed/pushed.

petrelharp commented:
Let me know when/if you want me to take another read through.

daniel-goldstein commented:
@petrelharp Ready whenever you can take a look!

petrelharp commented on docs/combinatorics.rst:
```diff
@@ -1,6 +1,210 @@
+elements. The order in which we select partitions of :math:`n` is
+determined by the efficient
+`rule_asc <http://jeromekelleher.net/generating-integer-partitions.html>`_
+algorithm for generating them.
```
Why not insert your nice explanation from before of the order that the integer partitions come in?  (I think you had such a description?)

petrelharp commented on docs/combinatorics.rst:
```diff
@@ -1,6 +1,210 @@
+    if all combinations of labels were exhausted. To prevent redundant labellings
+    we impose a *canonical labelling*. In the case of two symmetrical subtrees,
+    the left subtree must receive the minimum label from the label set. Notice
+    how this is the case in the right subtrees above.
```
Gee, also, I think you had a good description of what the order on labellings is, before? It'd be nice to put that back in.

petrelharp commented:
Well, it's still missing a lot of the interesting technical detail that says exactly what the ordering is - what's your reasoning for leaving that out?

jeromekelleher commented:
Looks great @daniel-goldstein, let's go through it later.

jeromekelleher commented:
OK, let's get this merged. We can follow up with more detail later, but this is enough to give some information for now.
